BUG: Groupby(sort=False) with datetime-like Categorical raises ValueError

sinhrks opened the pull request:
Closes #10505.

sinhrks committed:
BUG: Groupby(sort=False) with datetime-like Categorical raises ValueError

jreback commented on pandas/core/groupby.py:
```diff
@@ -1950,7 +1950,12 @@ def __init__(self, index, grouper=None, obj=None, name=None, level=None,
+                    unique_codes = unique1d(self.grouper.codes)
+                    # do not coerce to np.array to make datetime-like work
+                    cat = self.grouper.categories.take(unique_codes)
+                    self.grouper = self.grouper.reorder_categories(cat)
```
lets make a method in categorical to do like this (and just call it here)

jreback replied:
better yet I think unique for a cat should just do it this way

sinhrks replied:
How about adding  ``return_type`` kw to allow ``Categorical.unique`` to return ``Categorical`` like above?

jreback replied:
I think unique should always return a Categorical

(like above) - wonder why we don't have that

sinhrks replied:
Understood. Will confirm whether the API change has any side effect.

sinhrks commented:
Fixing ``Categorical.unique`` to return unique ``Categorical``. Would like to confirm the expected behavior based on current impl.

- unordered category: values and categories are sorted by appearance order, excluding unnecessary ones.
- ordered category: values are sorted by appearance order, categories keeps existing order excluding unnecessary ones.

```
import pandas as pd
import numpy as np

# unordered
cat = pd.Categorical(["b","b","a","a"], categories=["a","b","c"])
cat.unique()
# [b, a]
# Categories (2, object): [b, a]

cat = pd.Categorical(["b", "b", np.nan, "a"], categories=["a","b","c"]) 
cat.unique()
# [b, NaN, a]
# Categories (2, object): [b, a]

# ordered
cat = pd.Categorical(["b","b","a","a"], categories=["a","b","c"], ordered=True)
cat.unique()
# [b, a]
# Categories (2, object): [a < b]

cat = pd.Categorical(["b", "b", np.nan, "a"], categories=["a","b","c"], ordered=True) 
cat.unique()
# [b, NaN, a]
# Categories (2, object): [a < b]
```

jreback commented:
the uniques should not have ``nan`` in them. Further I think the ``ordered=True`` should order according to the categories (e.g. the present ones), this might be a small API change.

sinhrks commented:
Ah, correct. Fixed and updated the example above.

I understand ``groupby`` result must follows the same rule as above. It means ``groupby(Categorical(ordered=True), sorted=False)`` will have the index sorted by the order of categories, was previously ordered by appearance of values (unsorted).

sinhrks commented:
> uniques should not have nan

I have misunderstand this with ``value_counts``. Currently, ``unique`` returns array including ``NaN`` both normal and categorical.

```
pd.Series([1., 2., 1., np.nan]).unique()
# array([  1.,   2.,  nan])
pd.Series([1., 2., 1., np.nan], dtype='category').unique()
# array([  1.,   2.,  nan])
```

So, the fixed categorical result should preserve ``NaN`` in codes and not in categories as below? 

```
>>> pd.Series([1., 2., 1., np.nan], dtype='category').unique()
# [1, 2, NaN]
# Categories (2, float64): [1, 2]
```

jreback commented:
lgtm. merge when ready.

sinhrks commented:
Thanks, let me merge this once rebase and gets green.
